bootstrap/src/main/java/com/facebook/airlift/bootstrap/Bootstrap.java

```diff
@@ -326,24 +327,26 @@ static SortedMap<String, String> replaceWithEnvironmentVariables(
     {
         SortedMap<String, String> results = new TreeMap<>();
         properties.forEach((key, value) -> {
-            Matcher result = ENV_PATTERN.matcher(value);
-            if (result.matches()) {
-                String variableName = result.group(1);
-                String envValue = environment.get(variableName);
-
-                if (envValue == null) {
-                    String errorMessage = String.format(
-                            "Configuration property `%s` references `%s`, an undefiled environment variable.",
-                            key,
-                            variableName);
-                    onError.accept(key, errorMessage);
+            // TODO: Replace with Matcher.replaceAll(Function) or replace StringBuffer with StringBuilder after upgrade to Java 9+
+            StringBuffer result = new StringBuffer();
+            ImmutableSet<String> undefined = ImmutableSet.of();
+            Matcher matcher = ENV_PATTERN.matcher(value);
+            while (matcher.find()) {
+                String variable = matcher.group(1);
+                String variableValue = environment.get(variable);
+                if (variableValue == null) {
+                    undefined = ImmutableSet.<String>builder().addAll(undefined).add(variable).build();
```

Comment:
```suggestion
                    undefined.add(variable);
```

Then instantiate outside of the loop.